battle-service/src/main/java/com/company/battle/controllers/GameController.java

```diff
@@ -1,43 +1,81 @@
 package com.company.battle.controllers;
 
+import com.company.battle.mappers.GameMapper;
+import com.company.battle.mappers.SettingsMapper;
+import com.company.battle.services.GameService;
 import com.company.common.dtos.GameDto;
 import com.company.common.dtos.SettingsDto;
+import com.company.common.models.GameEntity;
+import com.company.common.models.SettingsEntity;
+import jakarta.validation.Valid;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.web.bind.annotation.*;
 
 import java.util.List;
 import java.util.UUID;
+import java.util.stream.Collectors;
 
 @RestController
 @RequestMapping("api/v1/game")
+@RequiredArgsConstructor
 public class GameController {
 
+    @Autowired
+    private final GameService gameService;
+
+    private final GameMapper gameMapper = GameMapper.INSTANCE;
+
+    private final SettingsMapper settingsMapper = SettingsMapper.INSTANCE;
+
     @GetMapping("/{id}")
     public GameDto getGameById(@PathVariable UUID id) {
-        return null;
+        GameEntity gameEntity = gameService.getById(id);
+        return gameMapper.toDTO(gameEntity);
     }
 
-    @GetMapping("/{link}")
+    @GetMapping("by-link/{link}")
     public GameDto getGameByLink(@PathVariable String link) {
-        return null;
+        GameEntity gameEntity = gameService.getByLink(link);
+        return gameMapper.toDTO(gameEntity);
+    }
+
+    @GetMapping("/public-games")
+    public List<GameDto> getPublicGames() {
+        List<GameEntity> gameEntities = gameService.getPublicGames();
+        return gameEntities.stream()
+                .map(gameMapper::toDTO)
+                .collect(Collectors.toList());
     }
 
-    @GetMapping
-    public List<GameDto> getActiveGames() {
-        return null;
+    @GetMapping("/find-game")
+    public GameDto findGame(@Valid @RequestBody SettingsDto settingsDto) {
+        SettingsEntity settingsEntity = settingsMapper.toEntity(settingsDto);
```

Comment:
what was the intent to find a game by SettingsDto?
you may use find by id which will return single possible game
or you can use findBY SettingsDto which may return list of games which fit to specific settings

Reply:
This is a search for a created game with the same settings. That is, a person has entered the settings that he needs. For example: squareSize = 3, linesCountForWin = 3 and other. And then the search happens. If no one has created such a game yet, then null will return and instead of searching there will be a request to create a game. And if there are such games, then the one that was created the most time ago is taken

Reply:
at least there might be more than 1 game fit to the settings you have listed, while you are returning single Game Dto.
you should use appropriate naming to get the  most "recently" created game which fit to the search criteria.
or change strategy to search games.

i would create separate SearchGameRequestDto where enumerate all possible search criteria ,and respond with List<GameDto> to allow user to choose from the list of options, this search requestdto may be handled in different ways, for examp,e using jpa criteria builder, and not usage of prepared queue.

for now i would limit functionality by returning just list of games in PENDING status.

Reply:
Okay, I can rename the method. But the idea of this method is that when the user starts searching for a game by specific parameters (by parameters in the settings), you need to return any suitable game to him. And I don't think it's right to search for the most recent game because older games will hang. Now, on the contrary, I return one of the oldest games according to these parameters. And I don't think it's appropriate to return all games in pending status, because it makes no sense to return a 4*4 game if the user is looking for a 3*3 game

Reply:
i don't understand what are you talking about:)
what i want to say is that settingsdto is something which consist of predefined configuration, so as long as there is predefined configuration user will have limited number search result (it may 1 , or 1000) depending on how many people use the same settings to create theirs games.
it doesn't make sense to return "ANY" single game from api standpoint, due to it makes your api very limited,and not extandable. what if tomorrow requirement will tell you "return 2 games which fit to the search criteria"?
usually you design api for consumers, and consumers able to control own behavior based on your api.
if you need to return just 1 game - yor method may still return "list" but there should be "limit" request param which will control number of max results in response. so consumer may say "return me list of N games which best fit to my search criteria", today N may be equal to 1, and tomorrow it may equal to 100. but in this case only consumers will require to adopt businesss logic, and not you, as backend developer.